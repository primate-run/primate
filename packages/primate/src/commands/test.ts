import { tests, type Body, type MockedResponse } from "#test";
import build from "@primate/core/build";
import green from "@rcompat/cli/color/green";
import red from "@rcompat/cli/color/red";
import root from "@rcompat/package/root";
import entries from "@rcompat/record/entries";
import equals from "@rcompat/test/equals";
import includes from "@rcompat/test/includes";
import type Dictionary from "@rcompat/type/Dictionary";
import type MaybePromise from "@rcompat/type/MaybePromise";
import serve from "./serve.js";

const directory  = "test";

type Check<T> = () => MaybePromise<[boolean, T, T | null]>;

const fetch_options = { redirect: "manual" } as const;

const first_error = (left: string, right: string) => {
  const length = left.length > right.length ? right.length : left.length;

  for (let i = 0; i < length; i++) {
    if (left[i] !== right[i]) {
      return i;
    }
  }
};

export default async () => {
  await build("testing", "web");
  const app = (await serve()).default;

  const files = await (await root()).join(directory)
    .list(({ path }) => path.endsWith(".ts") || path.endsWith(".js"));

  // side effects
  await Promise.all(files.map(file => file.import()));

  for (const test of tests) {
    const response = await fetch(`${app.url}${test.route}`, fetch_options);

    const checks: Check<Body | number>[] = [];

    const mocked_response: MockedResponse = {
      status: {
        equals(status) {
          checks.push(() => {
            return [response.status === status, status, response.status];
          });
        },
      },
      body: {
        equals(expected: Body) {
          checks.push(async () => {
            const actual = await (typeof expected === "string"
              ? response.text()
              : response.json());
            return [equals(actual, expected), expected, actual];
          });
        },
        includes(expected: Body) {
          checks.push(async () => {
            const actual = await (typeof expected === "string"
              ? response.text()
              : response.json());
            const $expected = typeof expected === "string"
              ? expected.replaceAll("\n", "").replaceAll("  ", "")
              : expected;
            return [includes(actual, $expected), $expected, actual];
          });
        },
      },
      headers: {
        includes(expected: Dictionary<string>) {
          checks.push(() => {
            const actual = Object.fromEntries(response.headers.entries());
            const lowercased = entries(expected)
              .keymap(([key]) => key.toLowerCase()).get();
            return [includes(actual, lowercased), lowercased, actual];
          });
        },
        get(name: string) {
          const actual = response.headers.get(name);

          return {
            equals(expected: string) {
              checks.push(() => {
                return [equals(actual, expected), expected, actual];
              });
            },
            includes(expected: string) {
              checks.push(() => {
                return [includes(actual, expected), expected, actual];
              });
            },
          };
        },
      },
    };
    test.tester(mocked_response);

    const results = await Promise.all(checks.map(async check => {
      try {
        return await check();
      } catch (error) {
        return [
          false, "()", "test execution failed",
        ];
      }
    }));

    const failed = results.find(result => !result[0]);
    const verb = test.verb.toUpperCase();

    if (failed !== undefined) {
      console.log(red(`${verb} ${test.route}`));
      const expected = JSON.stringify(failed[1]);
      const actual = JSON.stringify(failed[2]);
      const n = first_error(expected, actual)!;
      console.log(`expected: ${expected.slice(0, n)}${green(expected[n])}${expected.slice(n+1)}`);
      console.log(`actual:   ${actual.slice(0, n)}${red(actual[n])}${actual.slice(n+1)}`);
    } else {
      console.log(green(`${verb} ${test.route}`));
    }
  }

  await app.stop();
};
