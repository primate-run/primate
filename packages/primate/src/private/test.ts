import verbs from "@primate/core/verbs";
import type Dictionary from "@rcompat/type/Dictionary";

export type Body = string | Dictionary<string> | Dictionary<string>[];

export type MockedResponse = {
  status: {
    equals(status: number): void;
  };
  body: {
    equals(body: Body): void;
    includes(body: Body): void;
  };
  headers: {
    includes(headers: Dictionary<string>): void;
    get(header: string): {
      equals(value: string): void;
      includes(value: string): void;
    };
  };
};

type Verb = typeof verbs[number];

export type Tester = (response: MockedResponse) => void;

export type Route = string;

type Test = {
  verb: Verb;
  route: Route;
  tester: Tester;
};

export const tests: Test[] = [];

export default {
  ...Object.fromEntries(verbs.map(verb =>
    [verb, (route: Route, tester: Tester) => {
      tests.push({ verb, route, tester });
    }],
  )),
} as { [K in Verb]: (path: Route, tester: Tester) => void };
