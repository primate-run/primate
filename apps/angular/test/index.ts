import test from "primate/test";
import FileRef from "@rcompat/fs/FileRef";

const expected = await FileRef.join(import.meta.dirname, "index.expected.html")
  .text();

test.get("/", response => {
  response.body.includes(expected);
  /*response.client.query("span").equals("0");
  response.client.click("#increment");
  response.client.query("span").equals("1");
  response.client.click("#increment");
  response.client.query("span").equals("2");
  response.client.click("#decrement").click("#decrement");
  response.client.query("span").equals("0");*/
});
