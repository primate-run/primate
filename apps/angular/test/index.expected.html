<body><!--nghm--><app-root ng-version="20.0.6" ngh="0" ng-server-context="analog"><post-index ngh="0"><h1>bar</h1><h3>1 says:</h3><button>-</button><span>0</span><button>+</button></post-index></app-root><script id="ng-state" type="application/json">{"__nghData__":[{}]}</script></body>
